### PR TITLE
Auto return

### PR DESCRIPTION
**Pull request by guipublic**
With this PR, prover does not need to provide the main return value and can instead write in the prover.toml:
`return = ""`

Note that the return parameter must still be present in prover.toml, but we can give an empty value instead using an empty string.
Generating the proof will also create the verifier.toml (it overwrites if already  existing) with the public inputs (including the return value if there is one) corresponding to the provided prover.toml.

Here is a bit of context:
The `main()` function of a noir program can return a public value, which is computed by this noir program.
However, this public value is also a public input for the ZK proof so this requires the prover to compute by itself the value and put it in the provel.toml in order to generate the proof.
But the computation itself can be delicate if the noir program is doing a lof of (e.g. cryptographic) stuff. On the other hand, generating a proof of execution is also executing the program so noir should have already computed this value and it should not be necessary for the prover to provide the main return value.
This follows the concerns expressed in issue 355 [https://github.com/noir-lang/noir/issues/355]

### CODE REVIEW

**Comment by jfecher** [comment]
So when we have `return = ""` in the prover.toml it is now expected to write the actual return value into the Verifier.toml and create the file if it doesn't currently exist, correct? Does it rewrite the prover.toml with the correct return value as well?

**Comment by guipublic** [comment]
In all cases (i.e with a return="" , or with return="5", or with no return at all), it now generates the verifier.toml, and overwrites existing file.
The correct value will be in the verifier.toml, if there is a return value, whether or not it is specified in the prover.toml
- The prover.toml is never changed.
- The verifier.toml is always generated during **prove**
- The verifier.toml is never modified during **verify**

**Review by jfecher — CHANGES_REQUESTED** [review]

File: crates/nargo/src/cli/prove_cmd.rs [hunk 1]
```diff
@@ -38,12 +42,22 @@ fn prove(proof_name: &str, show_ssa: bool) -> Result<(), CliError> {
...
-) -> Result<BTreeMap<Witness, FieldElement>, CliError> {
+    witness_map: &BTreeMap<String, InputValue>,
+) -> Result<(BTreeMap<Witness, FieldElement>, Option<Witness>), CliError> {
     let mut solved_witness = BTreeMap::new();
 
     let mut index = 0;
-
+    let mut return_witness = None;
+    let return_witness_len = if let Some(pos) =
+        abi.parameters.iter().position(|x| x.0 == NodeInterner::main_return_name())
```
[the remark this answers is not in the document]
Comment (guipublic) [reply crates/nargo/src/cli/prove_cmd.rs:52]: Thanks, it's done

File: crates/nargo/src/cli/prove_cmd.rs [hunk 2]
```diff
@@ -71,9 +85,24 @@ fn process_abi_with_input(
...
+                    NodeInterner::main_return_name(),
+                    "input value {} is not defined",
+                    param_name
+                );
+                return_witness = Some(Witness::new(index + WITNESS_OFFSET));
+                index += 1;
+                //XXX We do not support (yet) array of arrays
+                for _i in 1..return_witness_len {
+                    Witness::new(index + WITNESS_OFFSET);
+                    index += 1
```
[the remark this answers is not in the document]
Comment (guipublic) [reply crates/nargo/src/cli/prove_cmd.rs:100]: Indeed the new witness is useless

File: crates/nargo/src/cli/prove_cmd.rs [hunk 3]
```diff
@@ -119,6 +151,45 @@ pub fn compile_circuit_and_witness<P: AsRef<Path>>(
...
+            let v = &witness_map[&i.0];
+
+            let iv = if matches!(*v, InputValue::Undefined) {
+                let w_ret = w_ret.unwrap();
+                match &i.1 {
+                    AbiType::Array { length, .. } => {
+                        let mut return_values = Vec::new();
+                        for i in 0..*length {
+                            return_values.push(
+                                *solved_witness.get(&Witness::new(w_ret.0 + i as u32)).unwrap(),
```
[the remark this answers is not in the document]
Comment (guipublic) [reply crates/nargo/src/cli/prove_cmd.rs:176]: Done

**Comment by guipublic** [comment]
Undefined arrays returned from main does not work, so I return an error in that case.
I will fix this issue in a separate PR.

**Review by vezenovm — APPROVED** [review]

File: crates/nargo/src/cli/prove_cmd.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jfecher) [reply crates/nargo/src/cli/prove_cmd.rs]: Actually, can it be simplified further to `solved_witness[&Witness::new(w_ret.0 + i as u32)]`?

**Review by jfecher — APPROVED** [review]

**Pull request merged**